emu_base/pulser_adapter.py

```diff
@@ -137,49 +205,10 @@ def _extract_omega_delta_phi(
     )
     qubit_positions = _get_qubit_positions(sequence.register)
 
-    def perp_dist(pos: torch.Tensor, axis: torch.Tensor) -> Any:
-        return torch.linalg.vector_norm(pos - torch.vdot(pos, axis) * axis)
-
-    times_to_amp_factors: dict[int, torch.Tensor] = {}
-    for ch, ch_samples in samples.channel_samples.items():
-        ch_obj = samples._ch_objs[ch]
-        prop_dir = torch.tensor(
-            ch_obj.propagation_dir or [0.0, 1.0, 0.0], dtype=torch.float64
-        )
-        prop_dir /= prop_dir.norm()
-
-        # each channel has a noise on its laser amplitude
-        # we assume each channel has the same noise amplitude currently
-        # the hardware currently has only a global channel anyway
-        sigma_factor = (
-            1.0
-            if amp_sigma == 0.0
-            else torch.max(torch.tensor(0), random_gaussian((1,), 1.0, amp_sigma)).item()
-        )
-        for slot in ch_samples.slots:
-            factors = (
-                torch.tensor(
-                    [
-                        math.exp(-((perp_dist(x, prop_dir) / laser_waist) ** 2))
-                        for x in qubit_positions
-                    ]
-                )  # the lasers have a gaussian profile perpendicular to the propagation direction
-                if laser_waist and ch_obj.addressing == "Global"
-                else torch.ones(
-                    len(q_ids)
-                )  # but for a local channel, this does not matter
-            )
-
-            # add the amplitude noise for the targeted qubits
-            factors[[x in slot.targets for x in q_ids]] *= sigma_factor
-
-            for i in range(slot.ti, slot.tf):
-                if i in times_to_amp_factors:  # multiple local channels at the same time
-                    # pulser enforces that no two lasers target the same qubit simultaneously
-                    # so only a single factor will be != 1.0 for each qubit
-                    times_to_amp_factors[i] = factors * times_to_amp_factors[i]
-                else:
-                    times_to_amp_factors[i] = factors
+    times_to_amp_factors = _get_amp_factors(
+        samples, amp_sigma, laser_waist, qubit_positions, q_ids
+    )
+    doppler_offset = _get_delta_offset(len(q_ids), temperature)
```

Comment:
nitpicking on this important code style guideline

> Ideally, declare and define each variable close to where it’s used.

so this line could be moved below

Reply:
done